Plugins/Cuda/src/Seeding2/FindDublets.cu

```diff
@@ -0,0 +1,194 @@
+  assert(false);
+  return 0.0f;
+}
+
+/// Function specialisation for middle-bottom spacepoint pairs
+template <>
+__device__ float getCotTheta<BottomSP>(float middleZ, float bottomZ,
+                                       float deltaR) {
+  return (middleZ - bottomZ) / deltaR;
+}
+
+/// Function specialisation for middle-top spacepoint pairs
+template <>
+__device__ float getCotTheta<TopSP>(float middleZ, float topZ, float deltaR) {
+  return (topZ - middleZ) / deltaR;
+}
+
+/// @}
+
+/// Kernel performing the spacepoint dublet finding
+///
+/// @tparam SPType The "other" spacepoint type used in the call
+///         (@c ::BottomSP or @c TopSP)
+/// @param[in] nMiddleSPs The number of middle spacepoints in @c middleSPs
+/// @param[in] middleSPs Properties of all of the middle spacepoints
+/// @param[in] nOtherSPs The number of other (bottom or top) spacepoints in
+///            @c otherSPs
+/// @param[in] otherSPs Properties of all of the other (bottom or top)
+///            spacepoints
+/// @param[in] deltaRMin Configuration parameter from @c Acts::SeedfinderConfig
+/// @param[in] deltaRMax Configuration parameter from @c Acts::SeedfinderConfig
+/// @param[in] cotThetaMax Configuration parameter from
+///            @c Acts::SeedfinderConfig
+/// @param[in] collisionRegionMin Configuration parameter from
+///            @c Acts::SeedfinderConfig
+/// @param[in] collisionRegionMax Configuration parameter from
+///            @c Acts::SeedfinderConfig
+/// @param[out] dubletCounts 1-D array of the middle-other dublets found
+///             for each middle spacepoint
+/// @param[out] dublets 2-D matrix of size @c nMiddleSPs x @c nOtherSPs, holding
+///             the "other" spacepoint indices for the identified dublets
+///
+template <int SPType>
+__global__ void findDublets(std::size_t nMiddleSPs,
+                            const Details::SpacePoint* middleSPs,
+                            std::size_t nOtherSPs,
+                            const Details::SpacePoint* otherSPs,
+                            float deltaRMin, float deltaRMax, float cotThetaMax,
+                            float collisionRegionMin, float collisionRegionMax,
+                            unsigned int* dubletCounts, std::size_t* dublets) {
+  // Figure out which dublet the kernel operates on.
+  const std::size_t middleIndex = blockIdx.x * blockDim.x + threadIdx.x;
+  const std::size_t otherIndex = blockIdx.y * blockDim.y + threadIdx.y;
+
+  // If we're outside of bounds, stop here.
+  if ((middleIndex >= nMiddleSPs) || (otherIndex >= nOtherSPs)) {
+    return;
+  }
+
+  // Calculate variables used in the compatibility check.
+  const float deltaR = getDeltaR<SPType>(middleSPs[middleIndex].radius,
+                                         otherSPs[otherIndex].radius);
+  const float cotTheta = getCotTheta<SPType>(middleSPs[middleIndex].z,
+                                             otherSPs[otherIndex].z, deltaR);
```

Comment:
I guess you could just use `if constexpr` and remove the need for the template and specializations.

Reply:
Hmm... I'll give it a try... [CUDA 11](https://news.developer.nvidia.com/cuda-11-toolkit/) does support C++17 by now. So it may be able to swallow that formalism.

Though in this case I really don't think that using templates is too bad. Compared to everything that Acts does, this templating is very localised. Plus I would be a bit hesitant to make the code only compile with CUDA 11 at this point. (It's just a bit too new/untested at this point.)

Reply:
Ah no sure. Compile-perf wise the specialization should be totally fine. I personally think `if constexpr` reads more naturally, but it just my opinion. Totally fine keeping the templates.